packages/compass-collection/src/modules/tabs.ts

```diff
@@ -182,56 +183,80 @@ const reducer: Reducer<CollectionTabsState> = (
       tabs: newTabs,
     };
   }
+  if (action.type === CollectionTabsActions.CollectionRenamed) {
+    const { tabs, activeTabIndex } = action;
+
+    const activeTabId = tabs[activeTabIndex]?.id;
+    return {
+      ...state,
+      tabs,
+      activeTabId,
+    };
+  }
   return state;
 };
 
 const subtabChanged = (id: string, name: string) => {
   return { type: CollectionTabsActions.SubtabChanged, id, name };
 };
 
+function createNewTabInternal(
+  collectionMetadata: CollectionMetadata,
+  dispatch: any,
+  globalAppRegistry: AppRegistry,
+  dataService: DataService | null
+) {
+  const collectionTabRole = globalAppRegistry.getRole(
+    'CollectionTab.Content'
+  )?.[0];
+  if (!collectionTabRole || !collectionTabRole.configureStore) {
+    throw new Error(
+      "Can't open a colleciton tab if collection tab role is not registered"
+    );
+  }
+  if (!dataService) {
+    throw new Error(
+      "Can't open a collection tab while data service is not connected"
+    );
+  }
+  const localAppRegistry = new AppRegistry();
+  const store = collectionTabRole.configureStore({
+    dataService,
+    globalAppRegistry,
+    localAppRegistry,
+    ...collectionMetadata,
+  });
+  const component = React.createElement(collectionTabRole.component, {
+    store,
+  });
+  const tab: CollectionTab = {
+    id: new ObjectId().toHexString(),
+    selectedSubTabName: store.getState().currentTab,
+    namespace: collectionMetadata.namespace,
+    type: collectionMetadata.isTimeSeries
+      ? 'timeseries'
+      : collectionMetadata.isReadonly
+      ? 'view'
+      : 'collection',
+    localAppRegistry,
+    component,
+  };
+  localAppRegistry.on('subtab-changed', (name: string) => {
+    dispatch(subtabChanged(tab.id, name));
+  });
+  return tab;
+}
+
 const createNewTab = (
   collectionMetadata: CollectionMetadata
 ): CollectionTabsThunkAction<CollectionTab> => {
   return (dispatch, getState, { globalAppRegistry, dataService }) => {
-    const collectionTabRole = globalAppRegistry.getRole(
-      'CollectionTab.Content'
-    )?.[0];
-    if (!collectionTabRole || !collectionTabRole.configureStore) {
-      throw new Error(
-        "Can't open a colleciton tab if collection tab role is not registered"
-      );
-    }
-    if (!dataService) {
-      throw new Error(
-        "Can't open a collection tab while data service is not connected"
-      );
-    }
-    const localAppRegistry = new AppRegistry();
-    const store = collectionTabRole.configureStore({
-      dataService,
+    return createNewTabInternal(
```

Comment:
Not actually sure completely why we moved it to a separate method. This `createNewTab` is already a shared action that returns a tab when dispatched, it's a bit unconventional, but the cleaenst way for now that gives us access to thunk provided services. Seems to me like both open collection and the new action that you added can just continue using it unless I'm missing something

Reply:
Here's my perhaps (incorrect) understanding.  `createNewTab` is a thunk action that opens a new tab with the given collection metadata at the end of the current set of tabs.  When a user renames a collection, we want to "refresh" the tabs that they have open for the old collection to have fresh collection data.  The easiest way to refresh the tabs is to close and reopen them.

If we simply dispatch a `createNewTab` for every tab that needs to be refreshed, we'd also have to emit corresponding `closeTab` actions for the old tab, _and_ we'd need to then reorder the tabs so that each newly opened tab is in the correct tab location.

I refactored this logic into a separate function because in my `collectionRenamed` thunk, I map over all the currently opened tabs and replace any that match the old namespace with a new tab.  That refreshes each stale collection tab while preserving the ordering without requiring multiple create/close actions + reordering of tabs.